### DISCUSSION TITLE
Padding question

### DISCUSSION THREAD

**Discussion by yshym** [asmvik/yabai#804] [Q&A]
Can I enable following settings only if a desktop contains >=2 windows?
```
yabai -m config top_padding                  10
yabai -m config bottom_padding               10
yabai -m config left_padding                 10
yabai -m config right_padding                10
```

**Comment by rockyzhang24** [comment]
You can use `yabai -m query --windows --space | jq 'length'` to get the number of windows in the current space.
Then you can build a command to enable some setting if and only if the number of windows in the current space is larger than 1 (i.e., `>=2`):

```
[[ $(yabai -m query --windows --space | jq 'length') -gt 1 ]] && yabai -m config top_padding 10
```

When the number of windows is less than 2, remove the padding:
```
[[ $(yabai -m query --windows --space | jq 'length') -lt 2 ]] && yabai -m config top_padding 0
```

Next, you should use yabai signal to make it run automatically
```
yabai -m signal --add event=window_created action="[[ $(yabai -m query --windows --space | jq 'length') -gt 1 ]] && yabai -m config top_padding 10"
yabai -m signal --add event=window_destroyed action="[[ $(yabai -m query --windows --space | jq 'length') -lt 2 ]] && yabai -m config top_padding 0"
```

Probably you need more signals to make it robust because `window_created` and `window_destroyed` may not cover all the situations (for example, `window_moved` for window moved into this space). I just give you some hints. Hope it is helpful.

**Comment by yshym** [comment]
Yes, it was very helpful. Thank you. I am trying to achieve something similar to [`smart_gaps` in i3-gaps](https://github.com/Airblader/i3#smart-gaps)

**Comment by yshym** [comment]
Don't understand, why signals are not working

``` shell
# tile apps
yabai -m rule --add app="^Emacs$" sticky=off manage=on

# float apps
yabai -m rule --add app="^System Preferences$" sticky=on layer=above manage=off

# layout
yabai -m config layout bsp
yabai -m config window_gap 7

if [[ $(yabai -m query --windows --space | jq "length") == 1 ]]; then
   # padding for one window
   yabai -m config top_padding 0
   yabai -m config bottom_padding 0
   yabai -m config left_padding 0
   yabai -m config right_padding 0
else
   # padding for two or more windows
   yabai -m config top_padding 10
   yabai -m config bottom_padding 10
   yabai -m config left_padding 10
   yabai -m config right_padding 10
fi

# signals
yabai -m signal --add \
    event=window_created \
    action="[[ $(yabai -m query --windows --space | jq 'length') -gt 1 ]] \
        && yabai -m config top_padding 10 \
        && yabai -m config bottom_padding 10 \
        && yabai -m config left_padding 10 \
        && yabai -m config right_padding 10"
yabai -m signal --add \
    event=window_destroyed \
    action="[[ $(yabai -m query --windows --space | jq 'length') == 1 ]] \
        && yabai -m config top_padding 0 \
        && yabai -m config bottom_padding 0 \
        && yabai -m config left_padding 0 \
        && yabai -m config right_padding 0"

yabai -m config window_placement second_child

yabai -m config split_ratio 0.50
yabai -m config auto_balance off
```

**Comment by yshym** [comment]
The problem is we should apply different padding settings on different desktops depending on number of windows

**Comment by rockyzhang24** [comment]
In the `action` of yabai signal, you should escape `$`, so change it to `action="[[ \$(yabai -m query --windows --space | jq 'length') -gt 1 ]] ....`.

> The problem is we should apply different padding settings on different desktops depending on number of windows

You can iterate over all the spaces, and for each space, get the number of windows there and then set the padding.
For example:
```
yabai -m query --spaces | jq '.[].index' | xargs -I{} yabai -m query --windows --space {} | jq 'length'
```
will output the number of windows in each space. Next, you just need to set the paddings based on the number (for example, use bash `while` loop to read each number and do something). 

If you are not familiar with `jq`, you can read its manual first (https://stedolan.github.io/jq/manual/). But it is better to read  yabai asciidoc first to understand what yabai could do (https://github.com/koekeishiya/yabai/blob/master/doc/yabai.asciidoc#signal). Yabai combined with `jq` is really powerful. :)

**Comment by yshym** [comment]
For now have no idea, why the following code doesn't work. `padding_refresh_command` works if executed using `/ust/bin/env sh -c` but doesn't work if triggered by a signal
``` shell
padding_refresh_command="p=\$((\$(yabai -m query --windows --space | jq 'length') == 1 ? 0 : 10)) && \
    yabai -m config --space mouse top_padding \$p && \
    yabai -m config --space mouse bottom_padding \$p && \
    yabai -m config --space mouse left_padding \$p && \
    yabai -m config --space mouse right_padding \$p"

yabai -m signal --add event=window_created action="$padding_refresh_command"
yabai -m signal --add event=window_destroyed action="$padding_refresh_command"
yabai -m signal --add event=application_launched action="$padding_refresh_command"
yabai -m signal --add event=application_terminated action="$padding_refresh_command"
```

**Comment by rockyzhang24** [comment]
Put all your code into my `yabairc` and it worked well on my end.

**Comment by yshym** [comment]
Does it work after you run this code in your terminal? Cause for me it doesn't

**Comment by rockyzhang24** [comment]
I pasted all your code above to the end of my `yabairc`. Then I restarted the yabai service by `launchctl kickstart -k "gui/${UID}/homebrew.mxcl.yabai"`. It worked well. 
Try to check your other yabai setting code in your `yabairc`. If the code in some line fails, all the lines after that won't be executed. So for example, if line 2 has an error, and you put your code above to line 10, your code won't be run.

**Comment by yshym** [comment]
Ok, here is my `yabairc`, but padding code should work separately, isn't it? These are just shell commands

``` shell
# load scripting-additions
sudo yabai --load-sa
yabai -m signal --add event=dock_did_restart action="sudo yabai --load-sa"

# global settings
# yabai -m config mouse_follows_focus on
# yabai -m config focus_follows_mouse autoraise
yabai -m config window_placement second_child
yabai -m config window_shadow off
yabai -m config window_border on
yabai -m config window_border_width 5
yabai -m config active_window_border_color 0xffa35763
yabai -m config normal_window_border_color 0xff353c54
yabai -m config insert_feedback_color 0xaa7c5c9c
yabai -m config split_ratio 0.50
yabai -m config auto_balance on
yabai -m config window_gap 7

# mission-control desktop labels
yabai -m space 1 --label web
yabai -m space 2 --label code
yabai -m space 3 --label social
yabai -m space 4 --label mail
yabai -m space 5 --label media

# window rules
yabai -m rule --add app="^Firefox$" space=web
yabai -m rule --add app="^Emacs$" space=code manage=on
yabai -m rule --add app="^Telegram$" space=social
yabai -m rule --add app="^Slack$" space=social
yabai -m rule --add app="^Discord$" space=social
yabai -m rule --add app="^Spark$" space=mail manage=on
yabai -m rule --add app="^Transmission$" space=media
yabai -m rule --add app="^Spotify$" space=media manage=on
yabai -m rule --add app="^Finder$" title="(Co(py|nnect)|Move|Info|Pref)" manage=off
yabai -m rule --add app="^Spotlight$" layer=above manage=off
yabai -m rule --add app="^Stickies$" manage=off
yabai -m rule --add app="^System Preferences$" layer=above manage=off

# general space layput
yabai -m config layout bsp

# padding
window_numbers=$(yabai -m query --spaces \
    | jq '.[].index'  \
    | xargs -I{} yabai -m query --windows --space {} \
    | jq "length")

di=1
for wn in $window_numbers; do
    p=$((wn == 1 ? 0 : 10))

    yabai -m config --space $di top_padding $p
    yabai -m config --space $di bottom_padding $p
    yabai -m config --space $di left_padding $p
    yabai -m config --space $di right_padding $p

    di=$((di+1))
done

# padding signals
padding_refresh_command="p=\$((\$(yabai -m query --windows --space | jq 'length') == 1 ? 0 : 10)) && \
    yabai -m config --space mouse top_padding \$p && \
    yabai -m config --space mouse bottom_padding \$p && \
    yabai -m config --space mouse left_padding \$p && \
    yabai -m config --space mouse right_padding \$p"

yabai -m signal --add event=window_created action="$padding_refresh_command"
yabai -m signal --add event=window_destroyed action="$padding_refresh_command"
yabai -m signal --add event=application_launched action="$padding_refresh_command"
yabai -m signal --add event=application_terminated action="$padding_refresh_command"

# signals
yabai -m signal --add event=window_destroyed action="yabai -m query --windows --window &> /dev/null || yabai -m window --focus mouse"
yabai -m signal --add event=application_terminated action="yabai -m query --windows --window &> /dev/null || yabai -m window --focus mouse"
```

**Comment by rockyzhang24** [comment]
> but padding code should work separately, isn't it? These are just shell commands

Yep. But in `yabaric`, the padding code is a setting, not an execution, i.e., it tells yabai to execute a piece of shell scripts when an event is triggered. If the code before it has an error, this setting won't be evaluated, i.e., the signal won't be created. 

When debugging, try to simplify the code by just leaving the piece of code you want to debug on and delete (or comment out) other codes temporarily.  You could remove all the rules and other personal settings, just keep the `yabairc` build-in settings and your padding setting you want to test. Then see whether it works on your end.

**Comment by yshym** [comment]
Doesn't work in either cases, whatever i comment

**Comment by yshym** [comment]
If i try to set some simple signals from the terminal manually after evaluating yabairc, they work, so i thought the issue was in padding signals code, but currently don't understand, how it can work for you

**Comment by yshym** [comment]
Both initial padding setup and signals should work after you run this code in your terminal. For me it doesn't

``` shell
# initial setup
window_numbers=$(yabai -m query --spaces \
    | jq '.[].index'  \
    | xargs -I{} yabai -m query --windows --space {} \
    | jq "length")

di=1
for wn in $window_numbers; do
    p=$((wn == 1 ? 0 : 10))

    yabai -m config --space $di top_padding $p
    yabai -m config --space $di bottom_padding $p
    yabai -m config --space $di left_padding $p
    yabai -m config --space $di right_padding $p

    di=$((di+1))
done

# signals
padding_refresh_command="p=\$((\$(yabai -m query --windows --space | jq 'length') == 1 ? 0 : 10)) && \
    yabai -m config --space mouse top_padding \$p && \
    yabai -m config --space mouse bottom_padding \$p && \
    yabai -m config --space mouse left_padding \$p && \
    yabai -m config --space mouse right_padding \$p"

yabai -m signal --add event=window_created action="$padding_refresh_command"
yabai -m signal --add event=window_destroyed action="$padding_refresh_command"
yabai -m signal --add event=application_launched action="$padding_refresh_command"
yabai -m signal --add event=application_terminated action="$padding_refresh_command"
```

**Comment by yshym** [comment]
@yanzhang0219 `yabai -m query --windows --space | jq` returns nothing if called in a signal action

**Comment by rockyzhang24** [comment]
`jq` outputs the results matching the filters to stdout. You should give some filters to `jq`.

**Comment by yshym** [comment]
No, without arguments it just pretty-prints the JSON. And it returns nothing no matter you give any arguments or not, if in signal action

**Comment by rockyzhang24** [comment]
Oh.. Right. My bad. My brain stopped working. Sorry. 
`yabai -m query --windows --space | jq` outputs the pretty JSON to the stdout. If you put it into the signal action, yabai runs it but yabai won't communicate with the stdout, so you see nothing. You must ask yabai to do something instead of outputing someting to stdout. For example, try `yabai -m query --windows --space | jq | say` to make your laptop say the output.

**Comment by yshym** [comment]
If we can't query space info in action, my trick can not be implemented using yabai commands and jq, as far as i understand

**Comment by rockyzhang24** [comment]
> If we can't query space info in action, my trick can not be implemented using yabai commands and jq, as far as i understand 

We CAN query sapce in action, but we should do something using the results, not just output them into `stdout`. Understand? If you just output the results into `stdout`, nothing will happen, so it looks like yabai does not execute the action. 

For example:
`yabai -m query --windows --space | jq '.[].app' | say` will make your computer say the application name of each window. Try it. and don't forget to turn up the volume of your computer :)

**Comment by yshym** [comment]
I don't need to use stdout at all. Check this code (it is the code, which is executed in action). Space info output is piped into jq and then the value in stored in the variable. But this code is not going to work, because `space_info` and `number_of_windows` have no values. You talk about querying yabai manually in my terminal, but the thing is about code in signal action
``` shell
space_info="$(yabai -m query --windows --space | jq)"
number_of_windows="$(yabai -m query --windows --space | jq 'length')"
padding=10

[ "$number_of_windows" -eq 1 ] && padding=0

osascript -e "display notification \"number_of_windows: $number_of_windows, padding: $padding\""
osascript -e "display notification \"$space_info\""

yabai -m config --space mouse top_padding "$padding"
yabai -m config --space mouse bottom_padding "$padding"
yabai -m config --space mouse left_padding "$padding"
yabai -m config --space mouse right_padding "$padding"
```

**Comment by rockyzhang24** [comment]
In your terminal, run:
```
space_info="$(yabai -m query --windows --space | jq)"
echo $sapce_info
```
you will see the contents in `space_info`.

![image](https://user-images.githubusercontent.com/11582667/104105966-ee1dbc00-5266-11eb-9d3e-59819b99c09a.png)

**Comment by yshym** [comment]
Ok, that's what i am talking about. Now try to create a script `refresh.sh` with my code and use it in a signal action `sh <path-to-script>` or just `<path-to-script>` (`application_launched` or `application_terminated` f.e.). If you don't want to waste your time, i can already say, that you won't get space info from there

**Comment by rockyzhang24** [comment]
`space_info` is not global but local. It is only available in the script file. 
Please paste your `refresh.sh` and the signal code in your `yabairc` here.

**Comment by yshym** [comment]
Hasn't changed anything. The following code in `yabairc`
``` shell
padding_refresh_command="space_info=\$(yabai -m query --windows --space | jq)
number_of_windows=\$(yabai -m query --windows --space | jq 'length')
padding=10

echo \"space_info: \$space_info, padding: \$padding, number_of_windows: \$number_of_windows\" >> \"\$HOME/yabailogs.txt\"

[ \$number_of_windows -eq 1 ] && padding=0

yabai -m config --space mouse top_padding \$padding
yabai -m config --space mouse bottom_padding \$padding
yabai -m config --space mouse left_padding \$padding
yabai -m config --space mouse right_padding \$padding"

yabai -m signal --add event=application_launched action="$padding_refresh_command"
yabai -m signal --add event=application_terminated action="$padding_refresh_command"
```
and `yabailogs.txt` after testing:
``` txt
space_info: , padding: 10, number_of_windows: 
space_info: , padding: 10, number_of_windows: 
space_info: , padding: 10, number_of_windows: 
space_info: , padding: 10, number_of_windows: 
```

**Comment by rockyzhang24** [comment]
I separated your refresh related code into a single script file. I tested it and verified that it worked well. Details are described below for your reference. 

1. Create a separate script file,` ~/refresh.sh` with the code below:
```bash
#!/usr/bin/env bash

space_info=$(yabai -m query --windows --space | jq)
number_of_windows=$(yabai -m query --windows --space | jq 'length')

padding=10

echo "space_info: $space_info, padding: $padding, number_of_windows: $number_of_windows" >> "/Users/yanzhang/yabailogs.txt"

[ $number_of_windows -eq 1 ] && padding=0

yabai -m config --space mouse top_padding $padding
yabai -m config --space mouse bottom_padding $padding
yabai -m config --space mouse left_padding $padding
yabai -m config --space mouse right_padding $padding
```

2. Change it execution mode. Run below code in the terminal directly
``` bash
chmod +x ~/refresh.sh
```

3. In your yabai, add signals like below:
```bash
yabai -m signal --add event=window_created action="~/refresh.sh"
yabai -m signal --add event=indow_destroyed action="~/refresh.sh"
```

(I change the event from `application_launched`, `application_terminated` to `window_created`, `window_destroyed` because `application` related events are not that convenient for testing.)

Attached a short video below for demo. 

https://user-images.githubusercontent.com/11582667/104107034-6c319100-526e-11eb-8b66-ed1785d28cf8.mov

**Comment by yshym** [comment]
Funny, it isn't still working for me. The problem is the same

**Comment by yshym** [comment]
> So, when the program is running the code (2), space_info variable has nothing.

But the program is running this code only on action, so there shouldn't be any case, when `space_info` variable has nothing

Implementation directly in `yabairc` using command string and implementation using separate script should have almost nothing different. `(1)` and `(2)` are both just one assignment

**Comment by rockyzhang24** [comment]
oh....So sorry. I missed that double quotes. 1, 2, and 3 are together in one pair of quote. Sorry.
Anyway, I made it run so the code is correct and nothing wrong with yabai. I have no idea why you cannot make it work.

**Comment by yshym** [comment]
`(3)` was the signals part. `(1)` and `(2)` only were the assignment xD

**Comment by rockyzhang24** [comment]
Okay. I see. But that's not important.
It works on my end...

**Comment by yshym** [comment]
Quite strange. I still have no values for `space_info` and `number_of_windows` variables

**Comment by rockyzhang24** [comment]
Just now, I put all your code directly into my `yabairc`

![image](https://user-images.githubusercontent.com/11582667/104108637-32b35280-527b-11eb-8fe4-b7fb405d8d8f.png)

And **it still works**. 

https://user-images.githubusercontent.com/11582667/104108653-51194e00-527b-11eb-9820-d3b9bc7ca95c.mov

**Comment by yshym** [comment]
It is strange, that if i just check space info,
``` shell
padding_refresh_command="space_info=\$(yabai -m query --windows --space | jq)
echo \"space_info: \$space_info\" >> \"\$HOME/yabailogs.txt\""

yabai -m signal --add event=window_created action="$padding_refresh_command"
yabai -m signal --add event=window_destroyed action="$padding_refresh_command"
```
signal action is executed, but i see nothing
``` txt
space_info: 
space_info: 
space_info: 
space_info: 
```

**Comment by rockyzhang24** [comment]
Interesting!
Do you try to reboot your computer? Or uninstall yabai completely and re-install yabai of the dev version (https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(from-HEAD)). Do you fully disable the SIP?

**Comment by yshym** [comment]
Yea, i rebooted my computer. SIP is disabled. yabai is from brew. I'll try to install the dev version
```
~
❯ csrutil status
System Integrity Protection status: disabled.
```

**Comment by rockyzhang24** [comment]
Sigh 😞 
I have no idea.

**Comment by yshym** [comment]
It seems like i skipped `/private/etc/sudoers.d/yabai` part while installing for the first time. But i installed the dev version, added sudoers file (scripting addition is now loaded for sure) and rebooted. Still nothing

**Comment by yshym** [comment]
It might be an OS version issue. Installation requirement is 11.0.1. I use 11.2

**Comment by rockyzhang24** [comment]
Okay... 11.1 is supported as well, but not sure about 11.2. Normally for a new macOS version, yabai needs to update a byte patterns in the payload.c file. We have to wait for the adaption.

**Comment by kpritam** [comment]
> I separated your refresh related code into a single script file. I tested it and verified that it worked well. Details are described below for your reference.
> 
> 1. Create a separate script file,` ~/refresh.sh` with the code below:
> 
> ```shell
> #!/usr/bin/env bash
> 
> space_info=$(yabai -m query --windows --space | jq)
> number_of_windows=$(yabai -m query --windows --space | jq 'length')
> 
> padding=10
> 
> echo "space_info: $space_info, padding: $padding, number_of_windows: $number_of_windows" >> "/Users/yanzhang/yabailogs.txt"
> 
> [ $number_of_windows -eq 1 ] && padding=0
> 
> yabai -m config --space mouse top_padding $padding
> yabai -m config --space mouse bottom_padding $padding
> yabai -m config --space mouse left_padding $padding
> yabai -m config --space mouse right_padding $padding
> ```
> 
> 1. Change it execution mode. Run below code in the terminal directly
> 
> ```shell
> chmod +x ~/refresh.sh
> ```
> 
> 1. In your yabai, add signals like below:
> 
> ```shell
> yabai -m signal --add event=window_created action="~/refresh.sh"
> yabai -m signal --add event=indow_destroyed action="~/refresh.sh"
> ```
> 
> (I change the event from `application_launched`, `application_terminated` to `window_created`, `window_destroyed` because `application` related events are not that convenient for testing.)
> 
> Attached a short video below for demo.
> 
>  Screen.Recording.2021-01-09.at.11.32.21.AM.mov

This worked for me as well. I am on BigSur.

**Comment by yshym** [comment]
> This worked for me as well. I am on BigSur.

It may work on 11.1 and previous versions, but may not work on 11.2, I suppose

**Comment by asmvik** [comment]
The signal and query system are not affected by SIP and/or macOS version.

**Comment by jasonvarga** [comment]
I tried to set this up and ran into issues too. The script runs fine, but it looks like even if a window is hidden it still appears in the query. If you actually _quit_ the applications so you only have one left, it works.

You could try filtering out non-visible windows, but then there's this: https://github.com/koekeishiya/yabai/issues/616
